pkg/webhook/preflight/preflight.go

```diff
@@ -0,0 +1,210 @@
+	// such as an infrastructure API client.
+	Checker interface {
+		// Init returns the checks that should run for the cluster.
+		Init(ctx context.Context, client ctrlclient.Client, cluster *clusterv1.Cluster) []Check
+	}
+
+	// Check represents a single preflight check that can be run against a cluster.
+	// It has a Name method that returns the name of the check, and a Run method executes
+	// the check, and returns a CheckResult.
+	// The Name method is used to identify the check if Run fails to return a result, for
+	// example if it panics.
+	Check interface {
+		Name() string
+		Run(ctx context.Context) CheckResult
+	}
+
+	// CheckResult represents the result of a check.
+	// It contains the name of the check, a boolean indicating whether the check passed, an
+	// error boolean indicating whether there was an internal error running the check, and a
+	// list of causes for the failure. It also contains a list of warnings that were
+	// generated during the check.
+	CheckResult struct {
+		Allowed bool
+		Error   bool
+
+		Causes   []Cause
+		Warnings []string
+	}
+
+	// Cause represents a cause of a check failure. It contains a message and an optional
+	// field that the cause relates to. The field is used to indicate which part of the
+	// cluster configuration the cause relates to.
+	Cause struct {
+		Message string
+		Field   string
+	}
+)
+
+type WebhookHandler struct {
+	client   ctrlclient.Client
+	decoder  admission.Decoder
+	checkers []Checker
+}
+
+func New(client ctrlclient.Client, decoder admission.Decoder, checkers ...Checker) *WebhookHandler {
+	h := &WebhookHandler{
+		client:   client,
+		decoder:  decoder,
+		checkers: checkers,
+	}
+	return h
+}
+
+type namedResult struct {
+	Name string
+	CheckResult
+}
+
+func (h *WebhookHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation == admissionv1.Delete {
+		return admission.Allowed("")
+	}
+
+	cluster := &clusterv1.Cluster{}
+	err := h.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Checks run only for ClusterClass-based clusters.
+	if cluster.Spec.Topology == nil {
+		return admission.Allowed("")
+	}
+
+	resultsOrderedByCheckerAndCheck := run(ctx, h.client, cluster, h.checkers)
+
+	// Summarize the results.
+	resp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			Allowed: true,
+			Result: &metav1.Status{
+				Details: &metav1.StatusDetails{},
+			},
+		},
+	}
+	internalError := false
+	for _, results := range resultsOrderedByCheckerAndCheck {
+		for _, result := range results {
+			if result.Error {
+				internalError = true
+			}
+			if !result.Allowed {
+				resp.Allowed = false
+			}
+			for _, cause := range result.Causes {
+				resp.Result.Details.Causes = append(resp.Result.Details.Causes, metav1.StatusCause{
+					Type:    metav1.CauseType(fmt.Sprintf("FailedPreflight%s", result.Name)),
```

Comment:
I think it is better to predefine the possible preflight causes as metav1.CauseType constants, see those examples in the metav1 package, such as:

const (
	// CauseTypeFieldValueNotFound is used to report failure to find a requested value
	// (e.g. looking up an ID).
	CauseTypeFieldValueNotFound CauseType = "FieldValueNotFound"
        .........
}

Reply:
I definitely considered this, but we need a field to hold the check name, so that clients can identiffy the check, for example, to offer the option of skipping the check.

We have three available fields; `Message` to hold the explanation, `Field` to reference the configuration, e.g., the field that needs to be changed to make the check pass, and finally, `Type` to hold the check name.